### PR TITLE
Adding cloud scripting

### PR DESCRIPTION
**Pull request by danmcp**
Adding scripting to help with cloud instance creation, management and the development and testing of instructlab against cloud instances.  Currently allows for ec2 and ibm instance creation, stop, start, terminate, and ssh (including running commands).  As well as setting up a development environment with instructlab code, prereq packages, and the ability to sync changes to the repo on the remote instance.

Usage:
```bash
./cloud-instance.sh <cloud-type> <command> [options]
```

Example:

```bash
# Launch a new instance with your `EC2_INSTANCE_NAME` and `EC2_INSTANCE_TYPE` specified
# in config with the instance type including Nvidia GPU(s)
./cloud-instance.sh ec2 launch
# Clone instructlab onto the instance and setup the development environment
./cloud-instance.sh ec2 setup-rh-devenv
# Install nvidia drivers and reboot
./cloud-instance.sh ec2 install-rh-nvidia-drivers
./cloud-instance.sh ec2 ssh sudo reboot
# Install instructlab
./cloud-instance.sh ec2 pip-install-with-nvidia
# ssh to the instance
./cloud-instance.sh ec2 ssh
# Run commands on the instance through ssh and return the result
./cloud-instance.sh ec2 ssh ls -la
./cloud-instance.sh ec2 ssh "source instructlab/venv/bin/activate && ilab sysinfo"
# Sync your local git repo to the repo on the instance
./cloud-instance.sh ec2 sync
# Make changes in your local git without committing
# Sync your changes to the remote instance with a temporary commit
./cloud-instance.sh ec2 sync -c
# When you're done, stop the instance
./cloud-instance.sh ec2 stop
# While stopped and not yet pruned from your account, you can restart it if needed
./cloud-instance.sh ec2 start
# When you are done, you can delete the instance
./cloud-instance.sh ec2 terminate
```

### CODE REVIEW

File: scripts/cloud/cloud-instance
Comment (russellb) [inline scripts/cloud/cloud-instance:3]: This will let you override the location by setting the `CONFIG_LOCATION` env var.

```suggestion
CONFIG_LOCATION=$(CONFIG_LOCATION:-$HOME/instructlab/cloud-config}
```

Comment (danmcp) [reply scripts/cloud/cloud-instance]: Fixed, thanks!

File: scripts/infra/cloud-instance
Comment (leseb) [inline scripts/infra/cloud-instance:55]: Can we timeout this loop to avoid running forever?

File: scripts/infra/cloud-instance
Comment (leseb) [inline scripts/infra/cloud-instance:63]: Print instance ID?

**Review by leseb — COMMENTED** [review]
I'm failing to understand why this helper script is needed in this repository? Can you help me? Thanks!

File: scripts/infra/cloud-instance
Comment (danmcp) [reply scripts/infra/cloud-instance:63]: Part of the intention of this utility is to not have to deal with the instance id.  Printing it out might make the user think it was needed for something.

**Comment by danmcp** [comment]
> I'm failing to understand why this helper script is needed in this repository? Can you help me? Thanks!

@leseb Perhaps there is a better location.  But the intention is to create a set of useful tooling to make it easy to script the launching of an instance and run instructlab.

File: scripts/infra/cloud-instance
Comment (danmcp) [reply scripts/infra/cloud-instance]: Thanks for the suggestion!

File: scripts/infra/cloud-instance
Comment (danmcp) [reply scripts/infra/cloud-instance]: Changed it to print the name and id next to it in ()

**Comment by leseb** [comment]
> > I'm failing to understand why this helper script is needed in this repository? Can you help me? Thanks!
> 
> @leseb Perhaps there is a better location. But the intention is to create a set of useful tooling to make it easy to script the launching of an instance and run instructlab.

I see the potential usefulness of this script for users with AWS accounts looking to experiment. However, I have some concerns about adding it to this repository. Accepting it might open the door to contributions for other cloud providers and local virtualized environments, which could create maintenance challenges. Everyone's workflow is unique, and it will be difficult to accommodate all variations. I'm also worried about setting a precedent without clear limits on what we accept. Additionally, the script appears to be compatible only with Fedora, which could cause issues for users on Debian-based distributions.

I'd like to hear what other maintainers think about this too. Thanks!

**Comment by danmcp** [comment]
> I see the potential usefulness of this script for users with AWS accounts looking to experiment. However, I have some concerns about adding it to this repository. Accepting it might open the door to contributions for other cloud providers and local virtualized environments, which could create maintenance challenges. Everyone's workflow is unique, and it will be difficult to accommodate all variations. I'm also worried about setting a precedent without clear limits on what we accept. Additionally, the script appears to be compatible only with Fedora, which could cause issues for users on Debian-based distributions.

The script does call out it's developed for Red Hat based systems in a way that would allow someone to add another distro later.  Note that as of today, the InstructLab READMEs also don't support anything other than Red Hat based linux.

As far as clear limits, this isn't core InstructLab code and could be considered best effort and only supporting an "easy path".

I don't have a strong opinion that this work must go into instructlab.  The reason I have started here is it seems like it could be helpful for others working on the project.

**Comment by russellb** [comment]
> > > I'm failing to understand why this helper script is needed in this repository? Can you help me? Thanks!
> > 
> > 
> > @leseb Perhaps there is a better location. But the intention is to create a set of useful tooling to make it easy to script the launching of an instance and run instructlab.
> 
> I see the potential usefulness of this script for users with AWS accounts looking to experiment. However, I have some concerns about adding it to this repository. Accepting it might open the door to contributions for other cloud providers and local virtualized environments, which could create maintenance challenges. Everyone's workflow is unique, and it will be difficult to accommodate all variations. I'm also worried about setting a precedent without clear limits on what we accept. Additionally, the script appears to be compatible only with Fedora, which could cause issues for users on Debian-based distributions.
> 
> I'd like to hear what other maintainers think about this too. Thanks!

I'm happy having it here. Nobody is required to use it, but it's a way to collaborate for those that use it.  If it gets out of hand, we can always move it later.  I don't have any better ideas for a place to share dev env scripts right now.

I don't think it should have to support all possible target dev environments out of the gate to go in. It just needs to be helpful to multiple people.

**Comment by nathan-weinberg** [comment]
Just to chime in, I'm fine with these scripts living here as well.

**Comment by danmcp** [comment]
Noting that I've got everything to a functional starting point for ec2 and ibm and the code is ready for any additional reviews/feedback.

**Review by leseb — CHANGES_REQUESTED** [review]

File: scripts/infra/cloud-instance
Comment (danmcp) [reply scripts/infra/cloud-instance]: This will work but will need to be in quotes so the || applies to the ssh.

File: scripts/infra/cloud-instance
Comment (danmcp) [reply scripts/infra/cloud-instance:304]: Thanks!

File: scripts/infra/cloud-instance
Comment (leseb) [reply scripts/infra/cloud-instance]: Good point, I overlooked that. Including everything into one line makes the fallback easier to read :)

File: scripts/infra/cloud-instance
Comment (danmcp) [reply scripts/infra/cloud-instance]: Fixed. Thanks!

File: scripts/infra/cloud-instance
Comment (danmcp) [reply scripts/infra/cloud-instance]: I've changed this to use -t to avoid finding anything other than a function.

File: scripts/infra/cloud-instance
Comment (danmcp) [reply scripts/infra/cloud-instance]: Same resolution as above.

**Review by leseb — COMMENTED** [review]
I'd like to hold this until https://github.com/instructlab/instructlab/pull/1211 is merged since it enables the shellcheck linter. Once merged, it's +1 for me since other maintainers agreed to have the script.

Thanks for addressing all the comments, could you in PR description add a workflow example of the script usage? So we can see a workflow example of using the script? Thanks!

**Comment by danmcp** [comment]
> I'd like to hold this until #1211 is merged since it enables the shellcheck linter. Once merged, it's +1 for me since other maintainers agreed to have the script.

Thanks for the review!

> 
> Thanks for addressing all the comments, could you in PR description add a workflow example of the script usage? So we can see a workflow example of using the script? Thanks!

Done!

**Comment by leseb** [comment]
@Mergifyio rebase

**Comment by danmcp** [comment]
> Can you rename `scripts/infra/cloud-instance` to `scripts/infra/cloud-instance.sh` so that the CI can pick it up? Thanks!

Done

File: scripts/infra/instructlab-cloud-config-example
Comment (leseb) [inline scripts/infra/instructlab-cloud-config-example:1]: The specs look relatively small depending on what we want to do. Can we add a section to guide users on choosing the optimal cloud instance type for their usage? It looks m4.large does not have any GPU, we could recommend a type that can take advantage of that?

File: scripts/infra/cloud-instance.sh
Comment (leseb) [inline scripts/infra/cloud-instance.sh:1]:
```suggestion
# SPDX-License-Identifier: Apache-2.0

#!/usr/bin/env -S bash -e
```

File: scripts/infra/README.md
Comment (danmcp) [reply scripts/infra/README.md]: Done

File: scripts/infra/README.md
Comment (danmcp) [reply scripts/infra/README.md]: Done

File: scripts/infra/README.md
Comment (danmcp) [reply scripts/infra/README.md]: Done

Comment (danmcp) [reply scripts/infra/README.md]: Examples provided.

File: scripts/infra/cloud-instance.sh
Comment (danmcp) [reply scripts/infra/cloud-instance.sh]: Note the shebang has to be the first line.

File: scripts/infra/instructlab-cloud-config-example
Comment (danmcp) [reply scripts/infra/instructlab-cloud-config-example:1]: I added additional instance type recommendations for EC2 and IBM

File: scripts/infra/README.md
Comment (nathan-weinberg) [inline scripts/infra/README.md:7]:
```suggestion
You'll need to setup the respective cloud CLI before using [cloud-instance.sh](cloud-instance.sh)
```

File: scripts/infra/README.md
Comment (nathan-weinberg) [inline scripts/infra/README.md:18]:
```suggestion
- A floating IP created within IBM Cloud. Config defaults to `whoami` for floating IP name.
```

File: scripts/infra/README.md
Comment (nathan-weinberg) [inline scripts/infra/README.md:27]:
```suggestion
You'll need to specify the `cloud-instance.sh` config before getting started. [instructlab-cloud-config-example](instructlab-cloud-config-example) is provided as a template. By default, you'll need to place your config in `~/.instructlab/cloud-config`, or you can specify the location as the environment variable `INSTRUCTLAB_CLOUD_CONFIG`
```

File: scripts/infra/README.md
Comment (nathan-weinberg) [inline scripts/infra/README.md:47]:
```suggestion
`./cloud-instance.sh <cloud-type> <command> [options]`
```

File: scripts/infra/README.md
Comment (danmcp) [reply scripts/infra/README.md]: Fixed, thanks!

File: scripts/infra/README.md
Comment (danmcp) [reply scripts/infra/README.md]: Fixed, thanks!

File: scripts/infra/README.md
Comment (danmcp) [reply scripts/infra/README.md]: Fixed, thanks!

**Review by leseb — APPROVED** [review]
I could add more comments in the README, but you've already addressed a lot, and I'm satisfied with the current state. I don't want to make it seem like I'm being overly critical of your contribution. We can always update the README.md again if needed. Thank you for your patience and great work!

**Review by nathan-weinberg — DISMISSED** [review]

**Comment by leseb** [comment]
@Mergifyio rebase

**Comment by leseb** [comment]
Mergify is not merging because of https://github.com/instructlab/instructlab/pull/1296

File: scripts/infra/README.md
Comment (russellb) [inline scripts/infra/README.md:70]:
```suggestion
./cloud-instance.sh ec2 sync -c
# When you're done, stop your instance
./cloud-instanfce.sh ec2 stop
# While stopped and not yet pruned from your account, you can restart it if needed
./cloud-instance.sh ec2 start
```

**Review by russellb — COMMENTED** [review]
This is great! I used it for my entire dev workflow on this PR: https://github.com/instructlab/instructlab/pull/1297

It worked perfectly!

I only have very minor suggestions, and then I'd like to merge it. Thank you!

File: scripts/infra/README.md
Comment (russellb) [reply scripts/infra/README.md:61]: to be clear, I don't think we need to block on this before merging, because the code is doing what our docs say ... can sort it out later

**Comment by leseb** [comment]
Putting on hold so that @russellb can be addressed since this is going to be merged once https://github.com/instructlab/instructlab/pull/1296 is merged.

File: scripts/infra/cloud-instance.sh
Comment (alimaredia) [inline scripts/infra/cloud-instance.sh:298]: FYI this is a more comprehensive list of packages I installed to get vllm working as a serving backend. 
```shell
sudo dnf install -y cuda-toolkit nvidia-gds libnccl-devel libcudnn8 nvidia-driver-NVML nvidia-driver-cuda-libs`
```

I can open up another PR with this if you'd rather not include it in this one.

File: scripts/infra/README.md
Comment (danmcp) [reply scripts/infra/README.md:71]: Suggestion accepted and I added terminate as well.  Thanks!

File: scripts/infra/instructlab-cloud-config-example
Comment (danmcp) [reply scripts/infra/instructlab-cloud-config-example]: Fixed, thanks!

File: scripts/infra/README.md
Comment (danmcp) [reply scripts/infra/README.md]: Fixed, thanks!

File: scripts/infra/cloud-instance.sh
Comment (danmcp) [reply scripts/infra/cloud-instance.sh]: I like the suggestion.  Going with it.  Changed for AMD as well.

File: scripts/infra/cloud-instance.sh
Comment (danmcp) [reply scripts/infra/cloud-instance.sh]: I went ahead and included.  Thanks!

**Comment by danmcp** [comment]
> This is great! I used it for my entire dev workflow on this PR: #1297
> 
> It worked perfectly!
> 
> I only have very minor suggestions, and then I'd like to merge it. Thank you!

@russellb Thanks for trying it out!

**Review by russellb — APPROVED** [review]
thank you!

**Review by leseb — APPROVED** [review]
Thanks!

**Review by nathan-weinberg — APPROVED** [review]

**Comment by russellb** [comment]
> @russellb @leseb seems the automerging isn't happening due to the below image

It's not merging because the lint job didn't run. I'll try to fix this.

**Comment by russellb** [comment]
This will fix the auto-merge issue in the future: https://github.com/instructlab/instructlab/pull/1330

**Comment by russellb** [comment]
Since the auto-merge config was the only thing holding this up, I'm just going to manually merge it. Thanks, Dan!

**Pull request merged**
